streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java

```diff
@@ -695,6 +695,58 @@ public void shouldShutdownTaskManagerOnCloseWithoutStart() {
         EasyMock.verify(taskManager);
     }
 
+    @Test
+    public void shouldNotAccessTaskManagerWhenPendingShutdownInRunOnce() {
+        final LogCaptureAppender appender = LogCaptureAppender.createAndRegister();
+
+        internalTopologyBuilder.addSource(null, "source", null, null, null, topic1);
+
+        final StreamThread streamThread = createStreamThread(clientId, new StreamsConfig(configProps(true)), true);
+
+        final Map<TaskId, Set<TopicPartition>> activeTasks = new HashMap<>();
+        final List<TopicPartition> assignedPartitions = new ArrayList<>();
+
+        // assign single partition
+        assignedPartitions.add(t1p1);
+        assignedPartitions.add(t1p2);
+        activeTasks.put(task1, Collections.singleton(t1p1));
+        activeTasks.put(task2, Collections.singleton(t1p2));
+
+        streamThread.taskManager().setAssignmentMetadata(activeTasks, Collections.emptyMap());
+        final MockConsumer<byte[], byte[]> mockConsumer = (MockConsumer<byte[], byte[]>) streamThread.consumer;
+        final Map<TopicPartition, Long> beginOffsets = new HashMap<>();
+        beginOffsets.put(t1p1, 0L);
+        beginOffsets.put(t1p2, 0L);
+        mockConsumer.updateBeginningOffsets(beginOffsets);
+
+        final Thread callbackThread = new Thread(() -> {
+            streamThread.rebalanceListener.onPartitionsRevoked(Collections.emptyList());
+            mockConsumer.assignForSubscribed(assignedPartitions);
+            streamThread.rebalanceListener.onPartitionsAssigned(assignedPartitions);
+        });
+
+        streamThread.setStateListener(
+            (t, newState, oldState) -> {
+                if (oldState == StreamThread.State.CREATED && newState == StreamThread.State.STARTING) {
+                    // Start triggering consumer callbacks to make sure we proceed to PARTITIONS_ASSIGNED stage.
+                    callbackThread.start();
+                } else if (oldState == StreamThread.State.PARTITIONS_REVOKED && newState == StreamThread.State.PARTITIONS_ASSIGNED) {
+                    // Immediately trigger shutdown call to switch the state.
+                    streamThread.shutdown();
```

Comment:
To test the race condition, should we call this when `newState == PARTITIONS_REVOKED`, ie, one state transition earlier?

Reply:
We couldn't reproduce the NPE in `partitionRevoked` because it's not expected to return records from `poll`. We need to trigger shutdown exactly during the call the `partitionAssigned` so that we are expecting records to return to attempt adding records to task manager. 
However current solution is not stable either, because it's not a correct way to trigger the condition. The natural way should leverage mock consumer to trigger assignment calls. Maybe we could talk in person and see how to reproduce it.